panther_driver/src/panther_kinematics.py

```diff
@@ -0,0 +1,173 @@
+#!/usr/bin/python3
+
+import math
+from numpy import clip
+from abc import abstractmethod
+
+import rospy
+
+from geometry_msgs.msg import Twist
+
+
+class PantherKinematics:
+    def __init__(self) -> None:
+        self.robot_width = 0.0
+        self.robot_length = 0.0
+        self.wheel_radius = 0.0  # Distance of the wheel center, to the roller center
+        self.encoder_resolution = None
+        self.gear_ratio = None
+        self.power_factor = None
+        self._wheels_enc_speed = [0.0, 0.0, 0.0, 0.0]
+        self.cmd_vel_command_time = rospy.Time.now()
+
+        self._lin_x = 0.0
+        self._lin_y = 0.0
+        self._ang_z = 0.0
+        self._robot_x_pos = 0.0
+        self._robot_y_pos = 0.0
+        self._robot_th_pos = 0.0
+        self._max_speed = 950.0
+        self._scale_factor_x = 0.25
+        self._scale_factor_y = 0.25
+        self._scale_factor_th = 0.125
+        self._const_factor = None
+        
+    @abstractmethod
+    def inverse_kinematics(self, data: Twist) -> None:
+        pass
+
+    @abstractmethod
+    def forward_kinematics(self, fl_ang_vel, fr_ang_vel, rl_ang_vel, rr_ang_vel, dt_) -> float:
+        pass
+
+    def _get_motor_speed(self, wheel_fl_ang_vel, wheel_fr_ang_vel, wheel_rl_ang_vel, wheel_rr_ang_vel) -> list:
+        # makes sure the rosparams are set correctly and calculates _const_factor only once
+        if self.power_factor is not None and self._const_factor is None:
+            self._const_factor = \
+                self.power_factor * float(self.encoder_resolution * self.gear_ratio) / (2.0 * math.pi)
+        elif self._const_factor is None:
+            return 0.0, 0.0, 0.0, 0.0
+        
+        # limit max power to 1000
+        fl_enc_speed = clip(self._const_factor * float(wheel_fl_ang_vel), -self._max_speed, self._max_speed)
+        fr_enc_speed = clip(self._const_factor * float(wheel_fr_ang_vel), -self._max_speed, self._max_speed)
+        rl_enc_speed = clip(self._const_factor * float(wheel_rl_ang_vel), -self._max_speed, self._max_speed)
+        rr_enc_speed = clip(self._const_factor * float(wheel_rr_ang_vel), -self._max_speed, self._max_speed)
```

Comment:
This can be replaced with lambda function in order to avoid using numpy. Numpy would be justified if those operations would be performed on arrays